### PR TITLE
ProcessExecutor: add teardown sequences

### PR DESCRIPTION
**Pull request by weissi**
This allows to write

```
        let p = ProcessExecutor(
            executable: "/usr/bin/swift",
            [ "build", "-c", "release" ],
            teardownSequence: [
                .sendSignal(SIGINT, allowedTimeToExitNS: 100_000_000), // SIGINT, then 100s
                .sendSignal(SIGTERM, allowedTimeToExitNS: 1_000_000_000), // SIGTERM, then 1s
                // and always SIGKILL if nothing worked
            ]
        )
```

which allows the process (`swift build` in this case) to clean up its child processes.

### CODE REVIEW

**Review by MaxDesiatov — APPROVED** [review]

File: Sources/AsyncProcess/ProcessExecutor.swift [hunk 1]
```diff
@@ -424,10 +532,12 @@ public final actor ProcessExecutor {
       )
       precondition(worked, "illegal running state \(original)")
 
-      if p.terminationReason == .uncaughtSignal {
-        terminationStreamProducer.yield(.signal(p.terminationStatus))
-      } else {
-        terminationStreamProducer.yield(.exit(p.terminationStatus))
+      for _ in 0..<2 {
```
Comment (kateinoigakukun) [inline Sources/AsyncProcess/ProcessExecutor.swift:535]: Could you elaborate on why we need to check the reason twice?

Comment (weissi) [reply Sources/AsyncProcess/ProcessExecutor.swift:535]: Yes, two tasks are waiting for the exit status:

1. the actual user
2. the termination sequence stuff that stops itself as soon as the process exits

Comment (kateinoigakukun) [reply Sources/AsyncProcess/ProcessExecutor.swift:535]: Ah, got it. `waitForChildToExit` is now awaited twice.

Comment (weissi) [reply Sources/AsyncProcess/ProcessExecutor.swift:535]: no

**Comment by kateinoigakukun** [comment]
Hmm, it seems the teardown sequence somehow does not wait for `allowedTimeNS` 🤔 

```
2024-04-05T22:46:26+0900 debug swift-async-process : arguments=["-c", "cat \"/var/folders/k2/q6yzck31253_q1j4kxkzq5jh0000gn/T/swift-sdk-generator-008142\"/*.pkg/Payload | gunzip -cd | (cd \"/var/folders/k2/q6yzck31253_q1j4kxkzq5jh0000gn/T/swift-sdk-generator-42E0B0\" && cpio -i)"] executable=/bin/sh pid=66892 [AsyncProcess] running command
^C2024-04-05T22:46:31+0900 debug swift-async-process : pid=66892 [AsyncProcess] we got cancelled
2024-04-05T22:46:31+0900 trace swift-async-process : pid=66892 step=sendSignal(2, allowedTimeNS: 100000000) [AsyncProcess] running teardown sequence
2024-04-05T22:46:31+0900 debug swift-async-process : outcome=processStillAlive pid=66892 step=sendSignal(2, allowedTimeNS: 100000000) [AsyncProcess] teardown sequence step complete
2024-04-05T22:46:31+0900 trace swift-async-process : pid=66892 step=sendSignal(15, allowedTimeNS: 1000000000) [AsyncProcess] running teardown sequence
2024-04-05T22:46:31+0900 debug swift-async-process : arguments=["-c", "cat \"/var/folders/k2/q6yzck31253_q1j4kxkzq5jh0000gn/T/swift-sdk-generator-008142\"/*.pkg/Payload | gunzip -cd | (cd \"/var/folders/k2/q6yzck31253_q1j4kxkzq5jh0000gn/T/swift-sdk-generator-42E0B0\" && cpio -i)"] executable=/bin/sh pid=66892 termination-reason=signal termination-status=15 [AsyncProcess] finished running command
2024-04-05T22:46:31+0900 debug swift-async-process : outcome=processHasExited pid=66892 step=sendSignal(15, allowedTimeNS: 1000000000) [AsyncProcess] teardown sequence step complete
SwiftSDKGenerator/Shell.swift:90: Fatal error
zsh: trace trap  swift run swift-sdk-generator make-linux-sdk
```

**Review by kateinoigakukun — APPROVED** [review]

**Comment by kateinoigakukun** [comment]
Ah, yes, 100s is `100_000_000_000`ns

**Pull request merged**